### PR TITLE
feat: add `RawEntryReader` and `OneshotWalEntryReader` trait

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4026 

## What's changed and what's your intention?

Add  `RawEntryReader` and `OneShotWalEntryReader` trait

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/wal/raw_entry_reader.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    RaftEngine(RegionId),
+    Kafka(String),
+}
+
+/// [RawEntryReader] provides the ability to read [RawEntry] from the underlying [LogStore].
+pub(crate) trait RawEntryReader: Send + Sync {
+    fn read<'a>(
+        &'a self,
+        ctx: &'a LogStoreReadCtx,
+        start_id: EntryId,
```
Comment (v0y4g3r) [inline src/mito2/src/wal/raw_entry_reader.rs:36]:
```suggestion
    fn read(
        &self,
        ctx: &LogStoreReadCtx,
        start_id: EntryId,
```

File: src/mito2/src/wal/raw_entry_reader.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+use store_api::storage::RegionId;
+
+use crate::error::Result;
+use crate::wal::EntryId;
+
+/// A stream that yields [RawEntry].
+pub type RawEntryStream<'a> = BoxStream<'a, Result<RawEntry>>;
+
+/// The context of [RawEntryReader].
+pub(crate) enum LogStoreReadCtx {
```
Comment (v0y4g3r) [inline src/mito2/src/wal/raw_entry_reader.rs:26]: Maybe rename it to namespace and remove those existing namespace design.

File: src/mito2/src/wal/wal_entry_reader.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use store_api::storage::RegionId;
+
+use super::raw_entry_reader::LogStoreReadCtx;
+use super::{EntryId, WalEntryStream};
```
Comment (evenyag) [inline src/mito2/src/wal/wal_entry_reader.rs:18]: `use crate::`

File: src/mito2/src/wal/raw_entry_reader.rs (hunk 2 above)
Comment (WenyXu) [reply src/mito2/src/wal/raw_entry_reader.rs]: I renamed it to `LogStoreNamespace`. Maybe remove the namespace in the next refactoring.

**Review by evenyag — APPROVED** [review]

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**